File: 剑指Offer/包含min函数的栈/minStack.py
class Solution:
    def __init__(self):
        self.arr = [];
        self.minL = [];
        self.mini = 2**31-1
    def push(self, node):
        # write code here
        self.arr.append(node)
        if node<=self.mini:
            self.minL.append(node)
            self.mini = node
        else:
            self.minL.append(self.mini)
    def pop(self):
        # write code here
        self.arr.pop()
        self.minL.pop()
        self.mini = self.minL[-1]

a = Solution()
a.push(3)
a.push(4)
a.push(2)
a.push(3)
a.pop()
a.pop()
a.pop()

File: 剑指Offer/包含min函数的栈/minStack.java
import java.util.Stack;
import java.util.ArrayList;

public class Solution {
    private ArrayList<Integer> arr = new ArrayList<Integer>();
    private ArrayList<Integer> minList = new ArrayList<Integer>();
    private int min = Integer.MAX_VALUE;
    
    public void push(int node) {
        arr.add(node);
        if(node<=min){
            minList.add(node);
            min = node;
        }else{
            minList.add(min);
        }
    }
    
    public void pop() {
        int end = arr.size()-1;
        arr.remove(end);
        minList.remove(end);
        min = minList.get(minList.size()-1);
    }
    
    public int top() {
        return arr.get(arr.size()-1);
    }
    
    public int min() {
        return min;
    }
}

import java.util.Stack;
import java.util.ArrayList;

public class Solution {
    private Stack<Integer> arr = new Stack<Integer>();
    private ArrayList<Integer> minList = new ArrayList<Integer>();
    private int min = Integer.MAX_VALUE;
    private int size = 0;
    
    public void push(int node) {
        size++;
        arr.push(node);
        if(node<=min){
            minList.add(node);
            min = node;
        }else{
            minList.add(min);
        }
    }
    
    public void pop() {
        arr.pop();
        minList.remove(size-1);
        size--;
        min = minList.get(size-1);
    }
    
    public int top() {
        int res = arr.pop();
        arr.push(res);
        return res;
    }
    
    public int min() {
        return min;
    }
}